### ISSUE TITLE
Problem with block_allocator for Cilk schedulers

### ISSUE THREAD

**Issue by neboat** [cmuparlay/parlaylib#18]
I'm getting sporadic failures (e.g., segfaults or incorrect results) when using a Cilk scheduler (e.g., OpenCilk) together with parlaylib's custom allocator within pbbsbench.  I'm able to reproduce the failure pretty consistently when running this command on an 8-core/16-hardware-thread AWS instance: `./runall -small -only comparisonSort/sampleSort`.  Unfortunately, I haven't been able to replicate the issue when running parlaylib's unit tests on the same machine.

I tracked down the problem to this line in `block_allocator.h`: https://github.com/cmuparlay/parlaylib/blob/99474c0a6613f6dbd944c5ffe2ab586a456fbf4b/include/parlay/internal/block_allocator.h#L193

The problem seems to be that `get_list()` may execute the `parallel_for` loop in `initialize_list()`.  When that happens under a parallel execution using a Cilk scheduler, the worker that starts executing `get_list()` is not guaranteed to be the same as the worker that returns from that call.  Hence, if, say, worker 1 calls `get_list()` and worker 2 returns from that call, then worker 2 will end up updating worker 1's `local_lists` structure.  Worker 1, meanwhile, might go and steal some other computation that calls `alloc()`, which can cause worker 1 to access its `local_lists` structure simultaneously with worker 2.  Safe to say, at that point, all sorts of terrible things might happen nondeterministically.

It's a bit easier to quickly trip the error by adding the following code after that line:
```cpp
 local_lists[id].head = get_list(); 
 if (id != worker_id()) {
   fprintf(stderr, "ERROR: Worker ID changed!\n");
   std::terminate();
 }
```

Let me know if you need anything else from me to examine this issue.

**Comment by DanielLiamAnderson** [comment]
Great find! Thanks for the very detailed investigation on your end. I've implemented a proposed fix at ae88af656f52e547a9951f38d940e68900c50fea.

On my end I was able to reproduce the issue by compiling pbbsbench with OpenCilk and running the sampleSort benchmark as you suggested. The proposed fix was successful on my end.

Could you try the patch and let me know if it corrects the issue for you? If it does, I can merge it.

**Comment by DanielLiamAnderson** [comment]
I've merged #19 which fixed the issue for me. Let me know if you still encounter the error.
